src/main/java/org/jenkinsci/plugins/workflow/actions/ArgumentsAction.java

```diff
@@ -0,0 +1,269 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.actions;
+
+import com.google.common.primitives.Primitives;
+import org.apache.commons.collections.CollectionUtils;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.graph.StepNode;
+import org.jenkinsci.plugins.workflow.steps.Step;
+import org.jenkinsci.plugins.workflow.steps.StepDescriptor;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Stores some or all of the arguments used to create and configure the {@link Step} executed by a {@link FlowNode}.
+ * This allows you to inspect information supplied in the pipeline script and otherwise discarded at runtime.
+ * Supplied argument values can be hidden and replaced with a {@link NotStoredReason} for security or performance.
+ *
+ * Important note: these APIs do not provide recursive guarantees that returned datastructures are immutable.
+ */
+public abstract class ArgumentsAction implements PersistentAction {
+
+    /** Used as a placeholder marker for {@link Step} arguments not stored for various reasons. */
+    public enum NotStoredReason {
+        /** Denotes an unsafe value that cannot be stored/displayed due to sensitive info. */
+        MASKED_VALUE,
+
+        /** Denotes an object that is too big to retain, such as strings exceeding {@link #MAX_RETAINED_LENGTH} */
+        OVERSIZE_VALUE
+    }
+
+    /** Largest String, Collection, or array size we'll retain -- provides a rough size limit on any single field.
+     *  Set to 0 or -1 to remove length limits.
+     */
+    private static final int MAX_RETAINED_LENGTH = Integer.getInteger(ArgumentsAction.class.getName()+".maxRetainedLength", 1024);;
+
+    /**
+     * Provides a basic check if an object contains any excessively large single elements.
+     * This recursively checks for String length, Map size, and array/collection size.
+     * We're not trying to cap total size just identify problem elements.
+     * @param o Object to check, with null allowed since we may see null inputs
+     * @param maxElements Max number of elements for a collection/map or characters in a string, or &lt; 0 to ignore length rules.
+     * @return True if object (or one of the contained objects) exceeds maxElements size.
+     */
+    public static boolean isOverSized(@CheckForNull Object o, final int maxElements) {
```

Comment:
> This would enable me to simplify the logic in workflow-cps

By making it more complex here I guess.

My point is that the decision about whether a given argument is reasonable and safe to store or not is potentially depend on the storage of the `FlowNode`. Under the current XML-based storage, defined in `workflow-support`, we need to be pretty careful. Under a possible future better engine, we could raise the limits, or take other factors into consideration. But it is the concrete subclass of `ArgumentsAction`—today in `workflow-cps`, really should be in `workflow-support` for architectural reasons (nothing specific to Groovy AFAICT)—and the code which attaches that action to a step node which is best placed to make that decision. Therefore it does not belong in the API.

(Even if there were no expectation that the policy might change in the future, it would be preferable to move it out of the API plugin anyway, since there is only a single implementation, and other plugins depending on `workflow-api` should not be calling these utilities: it is an implementation detail what is or is not masked or truncated.)